File: Hosts/Go/ToPython3/sanity/Main_test.go
package sanity

import (
	. "GoToPython3/testfuncs"
	"fmt"
	"os"
	"testing"
)

// --------------------------------------------------------------------
func TestMain(m *testing.M) {
	fmt.Println("Going to load TestFuncs module")
	MetaFFILoad("TestFuncs_MetaFFIGuest")
	fmt.Println("Start running")
	exitVal := m.Run()
	fmt.Println("Done running - going to free runtime")
	//Free()
	fmt.Println("Freed runtime")
	os.Exit(exitVal)
}

// --------------------------------------------------------------------
func TestHelloWorld(t *testing.T) {
	err := Hello_World()
	if err != nil {
		t.Fatal(err)
	}
}

// --------------------------------------------------------------------
func TestReturnsAnError(t *testing.T) {
	err := Returns_An_Error()
	if err == nil {
		t.Fatal("Error expected")
	}
}

// --------------------------------------------------------------------
func TestDivIntegers(t *testing.T) {
	res, err := Div_Integers(1, 2)
	if err != nil {
		t.Fatal(err)
	}

	if res != 0.5 {
		t.Fatalf("Expected 0.5, got: %v", res)
	}

	res, err = Div_Integers(1, 0)
	if err == nil {
		t.Fatal("Expected an error - divisor is 0")
	}
}

// --------------------------------------------------------------------
func TestJoinStrings(t *testing.T) {
	res, err := Join_Strings([]string{"A", "b", "C"})
	if err != nil {
		t.Fatal(err)
	}

	if res != "A,b,C" {
		t.Fatalf("Expected A,b,C. Got: %v", res)
	}
}

// --------------------------------------------------------------------
func TestWaitABit(t *testing.T) {

	fsec, err := Getfive_Seconds_MetaFFIGetter()
	if err != nil {
		t.Fatal(err)
	}

	shouldBeNullError, mffiError := Wait_A_Bit(fsec)
	if mffiError != nil {
		t.Fatal(mffiError)
	}

	if shouldBeNullError != nil {
		t.Fatal(shouldBeNullError)
	}
}

// --------------------------------------------------------------------
func TestTestMap(t *testing.T) {

	m, err := NewTestmap()
	if err != nil {
		t.Fatal(err)
	}

	_, err = m.Set("one", 1)
	if err != nil {
		t.Fatal(err)
	}

	one, err := m.Get("one")
	if err != nil {
		t.Fatal(err)
	}

	if one.(int64) != 1 {
		t.Fatalf("Expected one=1. one=%v", one)
	}

	err = m.Setname_MetaFFISetter("TheMap!")
	if err != nil {
		t.Fatal(err)
	}

	name, err := m.Getname_MetaFFIGetter()
	if err != nil {
		t.Fatal(err)
	}

	if name != "TheMap!" {
		t.Fatalf("Expected name=TheMap!. name=%v", name)
	}

	type User struct {
		ID string
	}

	u1 := User{ID: "TheUser!"}

	_, err = m.Set("user", u1)
	if err != nil {
		t.Fatal(err)
	}

	u2, err := m.Get("user")
	if err != nil {
		t.Fatal(err)
	}

	u2User, ok := u2.(User)
	if !ok {
		t.Fatalf("u2 is not of type User. u2=%v", u2)
	}

	if u1.ID != u2User.ID {
		t.Fatalf("user ID expeceted to be TheUser!. u2.ID=%v", u2User.ID)
	}
}

//--------------------------------------------------------------------

File: Hosts/Java/ToGroovy/sanity/Main_test.java
package sanity;

import openffi.*;

public class Main_test
{
	private static Test t = null;
	public static void main(String[] args)
	{
		try
		{
			t = new Test();

			testHelloWorld();
			testReturnsAnError();
			testDivIntegers();
			testJoinStrings();
		}
		catch(Exception e)
		{
			System.out.println("Unexpected exception has occurred during tests");
			e.printStackTrace();
			System.exit(1);
		}
	}

	private static void testHelloWorld() throws Exception
	{
		t.helloWorld();
	}

	private static void testReturnsAnError() throws Exception
	{
		try
		{
			t.returnsAnError();
			System.out.println("Test should have failed");
			System.exit(1);
		}
		catch(Exception e)
		{
			if(!e.getMessage().equals("Panic in Go function. Panic Data: An error from ReturnsAnError"))
			{
				System.out.println("Unexpected error message.\nExpected: Panic in Go function. Panic Data: An error from ReturnsAnError\nReceived: "+e.getMessage());
				System.exit(1);
			}
		}
	}

	private static void testDivIntegers() throws Exception
	{
		float res = t.divIntegers(1, 2);
        if(res != 0.5)
        {
            System.out.println("Expected 0.5, got: "+res);
            System.exit(1);
        }

        try
        {
            t.divIntegers(1, 0);
            System.out.println("Expected an error - divisor is 0");
            System.exit(1);
        }
        catch(Exception e)
        {
            if(!e.getMessage().equals("Panic in Go function. Panic Data: Divisor is 0"))
            {
                System.out.println("Unexpected error message.\nExpected: Panic in Go function. Panic Data: Divisor is 0\nReceived: "+e.getMessage());
                System.exit(1);
            }
        }
	}

	private static void testJoinStrings() throws Exception
	{
		String res = t.joinStrings(new String[]{"A","b","C"});
    	if(!res.equals("A,b,C"))
    	{
    	    System.out.println("Expected A,b,C. Got: "+res);
    	    System.exit(1);
    	}
	}
}